pkg/registry/stack.go

```diff
@@ -0,0 +1,137 @@
+package registry
+
+import (
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/openshift/ci-tools/pkg/api"
+)
+
+type stack struct {
+	records []stackRecord
+	partial bool
+}
+
+func stackForChain() stack {
+	return stack{partial: true}
+}
+
+func stackForWorkflow(name string, env api.TestEnvironment, deps api.TestDependencies) stack {
+	return stack{
+		records: []stackRecord{stackRecordForTest("workflow/"+name, env, deps)},
+		partial: true,
+	}
+}
+
+func stackForTest(name string, env api.TestEnvironment, deps api.TestDependencies) stack {
+	return stack{records: []stackRecord{stackRecordForTest("test/"+name, env, deps)}}
+}
+
+func (s *stack) push(r stackRecord) {
+	s.records = append(s.records, r)
+}
+
+func (s *stack) pop() {
+	s.records = s.records[:len(s.records)-1]
+}
+
+func (s *stack) errorf(format string, args ...interface{}) error {
+	var b strings.Builder
+	for i := range s.records {
+		b.WriteString(s.records[i].name)
+		b.WriteString(": ")
+	}
+	args = append([]interface{}{b.String()}, args...)
+	return fmt.Errorf("%s"+format, args...)
+}
+
+func (s *stack) resolve(name string) *string {
+	for _, r := range s.records {
+		for j, e := range r.env {
+			if e.Name == name {
+				for _, r := range s.records {
+					r.unusedEnv.Delete(e.Name)
+				}
+				return r.env[j].Default
+			}
+		}
+	}
+	return nil
+}
+
+func (s *stack) resolveDep(env string) string {
+	for _, r := range s.records {
+		for j, e := range r.deps {
+			if e.Env == env {
+				for _, r := range s.records {
+					r.unusedDeps.Delete(e.Env)
+				}
+				return r.deps[j].Name
+			}
+		}
+	}
+	return ""
+}
+
+func (s *stack) checkUnused(r *stackRecord) (ret []error) {
```

Comment:
The cubic loops here are horrible, but they're exceptional cases and the depth of the stack is never too big.